### PR TITLE
Core/Players: Improve rest state handling, allow several rest states to be active

### PR DESCRIPTION
**Pull request by Trisjdc**
Fixes https://github.com/TrinityCore/TrinityCore/issues/15325

### CODE REVIEW

**Comment by Rochet2** [comment]
+1 for making it a mask.
Also if the delay for unapplying areatrigger on leave zone is a thing, then maybe move the areatrigger unapply from `Player::Update` to update area and/or zone.

I think this PR atm makes a new issue with not getting the rest bonus.
The inn enter time is used to calculate the rest bonus and it should be cumulating when player is having the rested state, not when he is inside an areatrigger inn.

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -26588,3 +26554,19 @@ bool Player::ValidateAppearance(uint8 race, uint8 class_, uint8 gender, uint8 ha
...
+}
+
+void Player::RemoveRestFlag(RestFlag restFlag)
+{
+    _restFlagMask &= ~restFlag;
+    if (!_restFlagMask)
+        RemoveFlag(PLAYER_FLAGS, PLAYER_FLAGS_RESTING);
+
+    if (restFlag == REST_FLAG_IN_TAVERN)
+        InnEnter(time(nullptr), 0);
```
Comment (Rochet2) [inline src/server/game/Entities/Player/Player.cpp]: When entering a city the areatrigger applies the tavern flag, but it gets removed soon after when actually going further to the city.
At that point this refreshes the inn enter time so the time from entering the area trigger radius and leaving it is lost even though your rested state staid the same.

**Comment by Rochet2** [comment]
I think the timer should be changed so that when adding rested state and you have no previous state, the time is set to time(null). This way when you enter a new rested mask the rest time does not get reseted.

Then when removing rested state and you end up with no rested state the time should be set to 0 thus not able to gain rested bonus. Otherwise when removing flags your rested time should stay the same.

The areatrigger ID should be set separately when entering areatrigger radiuses.

**Comment by Trisjdc** [comment]
> Also if the delay for unapplying areatrigger on leave zone is a thing, then maybe move the areatrigger unapply from Player::Update to update area and/or zone.

It's not a thing, other cases look similar to Ironforge where the area has an appropriate flag.

>I think the timer should be changed so that when adding rested state and you have no previous state, the time is set to time(null). This way when you enter a new rested mask the rest time does not get reseted.

Yup, doing the change.

**Comment by Rochet2** [comment]
Nice work.

The rest bonus calculation looks really weird to me. Expecially that random 3% chance to actually get any resting bonus if you even are supposed to get any. Thats a separate issue though if even an issue.
https://github.com/Trisjdc/TrinityCore/commit/46b8df9a4986cefff55ab90918510fab4d32404a#diff-5589ee47b3a7ad220cff105a8fe7eac3L1726

**Comment by StormBytePP** [comment]
@Trisjdc better if you squash your code when you are finished so it is merged only in 1 commit.

**Comment by StormBytePP** [comment]
Will try it

**Comment by StormBytePP** [comment]
Depends on #15338

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -1723,15 +1722,20 @@ void Player::Update(uint32 p_time)
...
+            if (timeDiff >= 10)                             // freeze update
             {
-                float bubble = 0.125f*sWorld->getRate(RATE_REST_INGAME);
-                                                            // speed collect rest bonus (section/in hour)
-                SetRestBonus(GetRestBonus()+ time_inn*((float)GetUInt32Value(PLAYER_NEXT_LEVEL_XP)/72000)*bubble);
-                UpdateInnerTime(time(NULL));
+                _restTime = currTime;
+
+                float bubble = 0.125f * sWorld->getRate(RATE_REST_INGAME);
+                float extraPerSec = float(GetUInt32Value(PLAYER_NEXT_LEVEL_XP) / 72000) * bubble;
```
Comment (Rochet2) [inline src/server/game/Entities/Player/Player.cpp]: The float cast was removed. Need to add it back.

**Comment by StormBytePP** [comment]
@Trisjdc can you make the proposed float cast? So I can retest again and merge

**Pull request merged**